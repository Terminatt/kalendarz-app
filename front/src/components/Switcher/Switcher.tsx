import React, {
    useCallback, useEffect, useState,
} from 'react';
import { isDefined, isNumber } from '@utils/general';
import SwitcherLayout from './SwitcherLayout/SwitcherLayout';

export interface SwitcherOption<T> {
  label: string;
  value: T;
}
export interface SwitcherProps<T> {
  options: SwitcherOption<T>[];
  selected?: number | null;
  onChange?: (value: T | null, exceeds?: IndexExceed) => void;
}

export enum IndexExceed {
    NONE,
    LEFT,
    RIGHT,
}

const Switcher = <T, >(props: SwitcherProps<T>): React.ReactElement => {
    const [selectedIndex, setSelectedIndex] = useState<number | null>(null);
    const { selected, options, onChange } = props;
    const selectedItem = selected || selectedIndex;

    useEffect(() => {
        if (options.length === 0) {
            return;
        }
        setSelectedIndex(0);
    }, []);

    const onIndexChange = useCallback((index: number, exceeds?: IndexExceed): void => {
        if (onChange) {
            onChange(options[index] ? options[index].value : null, exceeds);
        }
    }, [onChange]);

    const onLeftButtonClick = useCallback(() => {
        if (!isNumber(selectedItem)) {
            return;
        }

        const newIndex = selectedItem - 1;

        if (newIndex < 0) {
            onIndexChange(newIndex, IndexExceed.LEFT);
            return;
        }

        onIndexChange(newIndex);

        if (isDefined(selected)) {
            return;
        }

        setSelectedIndex(newIndex);
    }, [selectedIndex, selected, onIndexChange]);

    const onRightButtonClick = useCallback(() => {
        if (!isNumber(selectedItem)) {
            return;
        }

        const newIndex = selectedItem + 1;

        if (newIndex > options.length - 1) {
            onIndexChange(newIndex, IndexExceed.RIGHT);
            return;
        }

        onIndexChange(newIndex);

        if (isDefined(selected)) {
            return;
        }

        setSelectedIndex(newIndex);
    }, [selectedIndex, selected, onIndexChange]);

    return (
        <SwitcherLayout onLeftClick={onLeftButtonClick} onRightClick={onRightButtonClick}>
            {selectedItem !== null ? options[selectedItem].label : ''}
        </SwitcherLayout>
    );
};

export default Switcher;
