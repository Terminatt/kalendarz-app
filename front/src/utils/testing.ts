export function matchMedia(): void {
    Object.defineProperty(window, 'matchMedia', {
        writable: true,
        value: jest.fn().mockImplementation((query: string) => ({
            matches: false,
            media: query,
            onchange: null,
            addListener: jest.fn(),
            removeListener: jest.fn(),
            addEventListener: jest.fn(),
            removeEventListener: jest.fn(),
            dispatchEvent: jest.fn(),
        })),
    });
}
