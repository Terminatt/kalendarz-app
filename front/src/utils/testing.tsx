import React, { ComponentType } from 'react';
import { render as rtlRender, RenderOptions } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';
import { RootState, rootReducer } from '@store/index';

export function matchMedia(): void {
    Object.defineProperty(window, 'matchMedia', {
        writable: true,
        value: jest.fn().mockImplementation((query: string) => ({
            matches: false,
            media: query,
            onchange: null,
            addListener: jest.fn(),
            removeListener: jest.fn(),
            addEventListener: jest.fn(),
            removeEventListener: jest.fn(),
            dispatchEvent: jest.fn(),
        })),
    });
}
export function isTestingEnv(): boolean {
    return process.env.JEST_WORKER_ID !== undefined || process.env.NODE_ENV === 'test';
}

export interface CustomOptions extends RenderOptions {
    preloadedState: RootState;
}

function render(
    ui: React.ReactElement,
    options?: CustomOptions,
) {
    const store = configureStore({ reducer: rootReducer, preloadedState: options?.preloadedState });
    function Wrapper({ children }: {children: React.ReactNode}) {
        return <Provider store={store}>{children}</Provider>;
    }

    return { element: rtlRender(ui, { wrapper: Wrapper as ComponentType, ...options }), store };
}

// re-export everything
export * from '@testing-library/react';
// override render method
export { render };
