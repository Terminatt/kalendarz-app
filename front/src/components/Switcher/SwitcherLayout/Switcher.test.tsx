import { fireEvent, render } from '@utils/testing';
import React from 'react';
import renderer from 'react-test-renderer';
import Switcher from '../Switcher';

const data = [
    {
        value: 1,
        label: 'Test1',
    },
    {
        value: 2,
        label: 'Test2',
    },
    {
        value: 3,
        label: 'Test3',
    },
    {
        value: 4,
        label: 'Test4',
    },
]
describe('Switcher Component', () => {
    it('matches the snapshot', () => {
        const tree = renderer
            .create(
                <Switcher options={[]} />
            )
            .toJSON();
        
        expect(tree).toMatchSnapshot();
    });

    it('does not trigger left click if there are no options', async () => {
        const  onChange = jest.fn();
        const { element } = render(
            <Switcher options={[]} onChange={onChange} />
        );

        fireEvent.click(await element.getByTestId('left-btn'));

        expect(onChange).not.toBeCalled();
    });

    it('does not trigger right click if there are no options', async () => {
        const  onChange = jest.fn();
        const { element } = render(
            <Switcher options={[]} onChange={onChange} />
        );

        fireEvent.click(await element.getByTestId('right-btn'));

        expect(onChange).not.toBeCalled();
    });

    it('triggers on left click when there are items', async () => {
        const  onChange = jest.fn();
        const { element } = render(
            <Switcher options={data} onChange={onChange} />
        );

        fireEvent.click(await element.getByTestId('left-btn'));

        expect(onChange).toBeCalledTimes(1);
    });

    it('triggers on right click when there are items', async () => {
        const  onChange = jest.fn();
        const { element } = render(
            <Switcher options={data} onChange={onChange} />
        );

        fireEvent.click(await element.getByTestId('right-btn'));

        expect(onChange).toBeCalledTimes(1);
    });

    it('triggers state change when there is no selected item', async () => {
        const  onChange = jest.fn();
        const { element } = render(
            <Switcher options={data} onChange={onChange} />
        );

        fireEvent.click(await element.getByTestId('right-btn'));
        fireEvent.click(await element.getByTestId('left-btn'));


        expect(onChange).toBeCalledTimes(2);
        expect(await element.queryByText(data[0].label)).not.toBeNull();
    });

    it('does not trigger state change whn there is selected item', async () => {
        const { element } = render(
            <Switcher selected={1} options={data} />
        );

        fireEvent.click(await element.getByTestId('left-btn'));


        expect(await element.queryByText(data[1].label)).not.toBeNull();
    });
});
