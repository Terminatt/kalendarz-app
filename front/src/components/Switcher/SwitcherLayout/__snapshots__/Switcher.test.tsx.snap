// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switcher Component matches the snapshot 1`] = `
<div
  className="switcher"
>
  <button
    className="ant-btn transparent-btn switcher-btn switcher-btn-left"
    data-testid="left-btn"
    onClick={[Function]}
    type="button"
  >
    <span
      aria-label="left"
      className="anticon anticon-left"
      role="img"
    >
      <svg
        aria-hidden="true"
        data-icon="left"
        fill="currentColor"
        focusable="false"
        height="1em"
        viewBox="64 64 896 896"
        width="1em"
      >
        <path
          d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 000 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
        />
      </svg>
    </span>
  </button>
  <span
    className="switcher-selected"
  >
    
  </span>
  <button
    className="ant-btn transparent-btn switcher-btn switcher-btn-right"
    data-testid="right-btn"
    onClick={[Function]}
    type="button"
  >
    <span
      aria-label="right"
      className="anticon anticon-right"
      role="img"
    >
      <svg
        aria-hidden="true"
        data-icon="right"
        fill="currentColor"
        focusable="false"
        height="1em"
        viewBox="64 64 896 896"
        width="1em"
      >
        <path
          d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
        />
      </svg>
    </span>
  </button>
</div>
`;
